Inlay hints support

nickysn opened the pull request:
This adds inlay hints support to nimsuggest. It adds a new command to nimsuggest, called 'inlayHints'.

Currently, it provides type information to 'var' and 'let' variables. In the future, inlay hints can also be added for 'const' and for function parameters. The protocol also reserves space for a tooltip field, which is not used, yet, but support for it can be added in the future, without further changing the protocol.

The change includes refactoring to allow the 'inlayHints' command to return a completely different structure, compared to the other nimsuggest commands. This will allow other future commands to have custom return types as well. All the previous commands return the same structure as before, so perfect backwards compatibility is maintained.

To use this feature, an update to the nim language server, as well as the VS code extension is needed.

Related PRs:
nimlangserver: https://github.com/nim-lang/langserver/pull/53
VS code extension: https://github.com/saem/vscode-nim/pull/134

nickysn committed:
initial (and still, very rough) attempt at adding inlay hints to nimsuggest

nickysn committed:
Store inside the SymInfoPair the value of the isDecl parameter, passed to the
suggestSym proc. Only show inlay suggest hints for symbols with isDecl = true.

nickysn committed:
Added nimsuggest-specific flag whether a declaration has a type specified
explicitly. We use this to avoid returning inlay type hints for vars that have
already a type, specified in code.

nickysn committed:
Changed the response to the new 'inlayHints' nimsuggest command to be a
different (custom) structure.

nickysn committed:
Added allowInsert flag to SuggestInlayHint

nickysn committed:
Added tooltip field to SuggestInlayHint (not used for now, but allows future
use, without changing the nimsuggest protocol for the result of the inlayHints
command)

nickysn commented:
CI failure is because nimlsp uses structures from the compiler sources. Suggest is now a base class, with multiple descendants, and all the members of what was previously Suggest are now in SuggestDef. Nimlsp needs to be updated in order to compile against the new compiler sources. Note that there's no change in the nimsuggest protocol, so already compiled versions of nimlsp should work with the new nimsuggest.

Araq commented:
I assume it's easier to keep backwards compatibility when you do away with the newly introduced inheritance? The added fields are nothing to worry about anyway.

nickysn committed:
Moved the fields of SuggestDef back into the Suggest base object, in order to
keep compatiblity with existing nimsuggest clients, that use the nimsuggest
structures from the compiler sources as a library.

nickysn committed:
Embed, instead of using inheritance for storing the inlay hints in the Suggest object

nickysn committed:
Proper implementation of the inRange function (also checks the column when the position is in the first or last line)

nickysn committed:
Disallow 'double click to insert' for inlay hints for 'for' variables, because
Nim doesn't support declaring a variable type for 'for' loops

nickysn commented:
Ok, got rid of the inheritance. The remaining CI failure looks like a random glitch and unrelated to my changes.

nickysn committed:
Fixed inlay hints location for variables, declared inside a template (they have
a `gensym appended to their name, internally)

nickysn committed:
Fixed potential range check error, introduced by the previous commit

Araq commented on compiler/suggest.nim:
```diff
@@ -105,7 +105,7 @@ proc getTokenLenFromSource(conf: ConfigRef; ident: string; info: TLineInfo): int
   elif ident[0] in linter.Letters and ident[^1] != '=':
     result = identLen(line, column)
-    if cmpIgnoreStyle(line[column..column + result - 1], ident) != 0:
+    if cmpIgnoreStyle(line[column..column + result - 1], ident[0..min(result-1,len(ident)-1)]) != 0:
```
Why?

nickysn replied:
Because ident may have extra stuff appended when part of a template. E.g. when a variable is part of a template, "`gensymXXX" is appended, where XXX are some numbers. In the previous version of the code, this would return a token length of 0. In turn, this causes the inlay hint to be put in the wrong place. Instead of:
```
template benchmark(benchmarkName: string, code: untyped) =
  block:
    debug "Started...", benchmark = benchmarkName
    let t0: float = epochTime()
    code
    let elapsed: float = epochTime() - t0
    let elapsedStr: string = elapsed.formatFloat(format = ffDecimal, precision = 3)
    debug "CPU Time", benchmark = benchmarkName, time = elapsedStr
```
You would get:
```
template benchmark(benchmarkName: string, code: untyped) =
  block:
    debug "Started...", benchmark = benchmarkName
    let : floatt0 = epochTime()
    code
    let : floatelapsed = epochTime() - t0
    let : stringelapsedStr = elapsed.formatFloat(format = ffDecimal, precision = 3)
    debug "CPU Time", benchmark = benchmarkName, time = elapsedStr
```

Araq replied:
Sucks, but ok.

Araq commented on compiler/suggest.nim:
```diff
@@ -175,58 +175,89 @@ proc symToSuggest*(g: ModuleGraph; s: PSym, isLocal: bool, section: IdeCmd, info
     result.line = toLinenumber(infox)
     result.column = toColumn(infox)
-    result.tokenLen = if section != ideHighlight:
+    result.tokenLen = if not (section in {ideHighlight, ideInlayHints}):
```
```suggestion
    result.tokenLen = if section notin {ideHighlight, ideInlayHints}:
```

Araq commented on compiler/suggest.nim:
```diff
@@ -175,58 +175,89 @@ proc symToSuggest*(g: ModuleGraph; s: PSym, isLocal: bool, section: IdeCmd, info
+      result.add($suggest.endCol)
+
+proc suggestToSuggestInlayHint*(sug: Suggest): SuggestInlayHint =
+  new(result)
+  result.kind = sihkType
+  result.line = sug.line
+  result.column = sug.column + sug.tokenLen
+  result.label = ": " & sug.forth
+  result.paddingLeft = false
+  result.paddingRight = false
+  result.allowInsert = true
+  result.tooltip = ""
```
```suggestion
  SuggestInlayHint(
    kind: sihkType,
    line: sug.line,
    column: sug.column + sug.tokenLen,
    label: ": " & sug.forth,
    paddingLeft: false,
    paddingRight: false,
    allowInsert: true,
    tooltip: ""
  )
```

Araq commented on nimsuggest/nimsuggest.nim:
```diff
@@ -1066,6 +1104,19 @@ proc executeNoHooksV3(cmd: IdeCmd, file: AbsoluteFile, dirtyfile: AbsoluteFile,
+    i += parseInt(tag, endCol, i)
+    let s = graph.findSymDataInRange(file, line, col, endLine, endCol)
+    for q in s:
+      if q.sym.kind in [skLet, skVar, skForVar] and q.isDecl and not q.sym.hasUserSpecifiedType:
```
```suggestion
      if q.sym.kind in {skLet, skVar, skForVar} and q.isDecl and not q.sym.hasUserSpecifiedType:
```

Araq committed:
Update compiler/suggest.nim

Araq committed:
Update nimsuggest/nimsuggest.nim

nickysn committed:
Update compiler/suggest.nim

Co-authored-by: Andreas Rumpf <rumpf_a@web.de>
